sotodlib/tod_ops/demodulation.py

```diff
@@ -0,0 +1,264 @@
+        _remove_demod_placeholders(aman)
+    except:
+        _remove_demod_placeholders(aman)
+        raise
+    return
+
+
+def get_phase_fit_signals(aman, middle_relative_time, index_limit=100,
+                          threshold=10., plot=False):
+    '''Get the signals we want to fit a phase to. These should have strong power
+       in the region we're interested in.
+
+    args:
+
+        aman                 - axis manager to use
+        middle_relative_time - (relative) time since the beginning of the
+                               timestream that we want to look around
+        index_limit          - +/- limits of indices around
+                               middle_relative_time that we look
+        threshold            - power threshold that we cut off when checking
+        plot                 - show plots of data
+
+    returns a list of signals as well as the times they correspond to.
+
+    '''
+
+    phase_fit_signals = []
+    if (plot):
+        plt.figure(figsize=(10, 5))
+
+    for i in range(0, aman.dets.count):
+        if aman.dets.vals[i] in aman.flags.has_cuts(['trends']):
+            continue
+        relative_time = aman.timestamps - aman.timestamps[0]
+        middle_time = np.where(relative_time < middle_relative_time)[0][-1]
+
+        start, stop = (middle_time - index_limit, middle_time + index_limit)
+
+        # We only take specific signals with power below the threshold
+        # to avoid weirdly saturated detectors
+        if np.max(np.abs(aman.signal[i][start: stop])) < threshold:
+            if (plot):
+                plt.plot(relative_time[start: stop],
+                         aman.signal[i][start: stop])
+
+            phase_fit_signals.append(aman.signal[i][start: stop])
+    if (plot):
+        plt.grid()
+        plt.show()
+
+    return phase_fit_signals, aman.timestamps[start: stop]
+
+
+def _fit_sin(tt, yy, freq=8, plot=False):
+    '''Fit sin to the input time sequence, and return fitting parameters
+    "amplitude", "omega", "phase", "offset", "freq"'''
+    # Create sine func to fit.
+    omega = 2 * np.pi * freq
+
+    def sinfunc(t, A, phi, c):
+        return np.abs(A) * np.sin(omega * t + phi) + c
+
+    tt = np.array(tt)
+    yy = np.array(yy)
+    # guess the amplitude and offset of this sine wave
+    guess_amp = np.std(yy) * 2.**0.5
+    guess_offset = np.mean(yy)
+    guess = np.array([guess_amp, 0., guess_offset])
+
+    # fit to the data.
+    popt, pcov = curve_fit(sinfunc, tt, yy, p0=guess, maxfev=10000)
+
+    A, phi, c = popt
+
+    if (plot):
+        plt.plot(tt, yy, label='data')
+        plt.plot(tt, sinfunc(tt, *popt), label='fit, phase = %.2f' % phi)
+        plt.legend()
+        plt.show()
+
+    return {"amplitude": A, "phase": phi, "offset": c, "freq": freq}
+
+
+def fit_phase(aman, middle_relative_time, index_limit=180, threshold=.8,
+              freq=8.0, plot=False):
+    '''Fit a sinewave to the data to find the chopper phase.
```

Comment:
Make this comment a bit more general? This can be used in more than chopper data.

Maybe "fit sinewave to modulated data, only useful in regions where modulation signal has sufficiency signal-to-noise".

Reply:
agreed, done